### PR TITLE
Add basic support for CUDA Graph

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
New features

### PR changes
APIs
### Describe
Add basic support for CUDA Graph, including:

- Memory pool for CUDA Graph capturing. CUDA Graph needs to cache all Tensor address without calling any `cudaFree`.
- Basic APIs like `CUDAGraph.capture_begin/capture_end/replay/reset`. Notice that this API is in the experimental stage, and may be changed in the future.
- Use `CUDAGraphCaptureModeGuard` to switch to `cudaStreamCaptureModeRelaxed` to skip the unsupported `cudaMalloc` during capturing.

Usage:
```
from paddle.device.cuda.graphs import CUDAGraph

input = ... # define the input tensor. The definition of the input tensor should be before the `CUDAGraph.capture_start()`.

graph = CUDAGraph()
graph.capture_start()
output = ... # do some GPU operations here
graph.capture_end()

for _ in range(BATCH_NUM):
    input.copy_(input_tensor, False) # input_tensor is the input data of the model, may be from DataLoader
    graph.replay()
    print(output)

graph.reset() # it is not required, but it is better to call this method to release cached memory asap
```

TODO: when using CUDA Graph,
- Cache CuDNN descriptor. Otherwise, errors would raise during capturing.
- Disable CuDNN exhaustive search. Otherwise, errors would raise during capturing.
- Disable `FLAGS_sync_all_reduce` when using distributed training. The `FLAGS_sync_all_reduce` would call `cudaStreamSynchronize`, which is not supported during capturing.
- Modify ParallelExecutor to support CUDA Graph.

### CODE REVIEW

File: paddle/fluid/memory/allocation/allocator_facade.cc [hunk 1]
```diff
@@ -279,12 +351,57 @@ class AllocatorFacadePrivate {
...
+    for (auto& item : allocator->allocators_) {
+      auto& old_allocator = item.second;
+      old_allocator = CUDAGraphAllocator::Create(old_allocator);
+    }
+    VLOG(10) << "Prepare memory pool for CUDA Graph with ID " << id;
+  }
+
+  void RemoveMemoryPoolOfCUDAGraph(CUDAGraphID id) {
+    auto iter = cuda_graph_allocator_map_.find(id);
+    PADDLE_ENFORCE_EQ(iter != cuda_graph_allocator_map_.end(), true,
```
Comment (zhiqiu) [inline paddle/fluid/memory/allocation/allocator_facade.cc]: Can use PADDLE_ENFORCE_NE  directly

File: paddle/fluid/platform/cuda_graph.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  static std::unique_ptr<CUDAGraph> capturing_graph_;
+};
+
+#if CUDA_VERSION >= 10010
+class CUDAGraphCaptureModeGuard {
+  DISABLE_COPY_AND_ASSIGN(CUDAGraphCaptureModeGuard);
+
+ public:
+  explicit CUDAGraphCaptureModeGuard(cudaStreamCaptureMode new_mode) {
+    old_mode_ = new_mode;
```
Comment (zhiqiu) [inline paddle/fluid/platform/cuda_graph.h]: why old_mode_ = new_mode?

File: paddle/fluid/platform/cuda_graph.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    PADDLE_ENFORCE_CUDA_SUCCESS(
+        cudaThreadExchangeStreamCaptureMode(&old_mode_));
+  }
+
+  ~CUDAGraphCaptureModeGuard() PADDLE_MAY_THROW {
+    PADDLE_ENFORCE_CUDA_SUCCESS(
```
Comment (zhiqiu) [inline paddle/fluid/platform/cuda_graph.h]: Is it ok to raise exception in the destructor?

File: paddle/fluid/platform/gpu_info.cc [hunk 4]
```diff
@@ -557,6 +558,7 @@ class RecordedCudaMallocHelper {
 #ifdef PADDLE_WITH_HIP
     auto result = hipMalloc(ptr, size);
 #else
+    CUDAGraphCaptureModeGuard capture_mode_guard{cudaStreamCaptureModeRelaxed};
```
Comment (zhiqiu) [inline paddle/fluid/platform/gpu_info.cc:561]: imho, call this func when IsCUDAGraphCapturing is true.

File: paddle/fluid/platform/cuda_graph.h (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/platform/cuda_graph.h]: Yes. Although it is not recommended to raise exception in the destructor, I think that we should not hide the exception. If exception is raised in the destructor, `std::terminate` would be called to stop the process immediately.

File: paddle/fluid/platform/gpu_info.cc (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/platform/gpu_info.cc:561]: Done.

File: paddle/fluid/memory/allocation/allocator_facade.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/memory/allocation/allocator_facade.cc]: Done.

File: paddle/fluid/platform/cuda_graph.h (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/platform/cuda_graph.h]: Change the variable name and add some comments for better understanding.

File: paddle/fluid/platform/cuda_graph.h (hunk 3 above)
Comment (zhiqiu) [reply paddle/fluid/platform/cuda_graph.h]: > Yes. Although it is not recommended to raise exception in the destructor, I think that we should not hide the exception. If exception is raised in the destructor, `std::terminate` would be called to stop the process immediately.

Maybe we need a `PADDLE_WARN_CUDA_SUCCESS`, etc.

Comment (sneaxiy) [reply paddle/fluid/platform/cuda_graph.h]: > > Yes. Although it is not recommended to raise exception in the destructor, I think that we should not hide the exception. If exception is raised in the destructor, `std::terminate` would be called to stop the process immediately.
> 
> Maybe we need a `PADDLE_WARN_CUDA_SUCCESS`, etc.

Hard to do this. Suppose that we have a common method `void func()`. The `func` may be called anywhere, inside destructor or outside destructor, but we have to only write one of `PADDLE_ENFORCE_CUDA_SUCCESS` and  `PADDLE_ENFORCE_WARN_SUCCESS` inside its implementation.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**